### PR TITLE
AudioEffectPitchShift: Fix distortion when pitch is 1.0

### LINKED ISSUE #96510: AudioEffectPitchShift has audible effect when pitch = 1
### Tested versions

- Reproducible in 4.3.stable

### System information

Godot v4.3.stable - Windows 10.0.19045 - GLES3 (Compatibility) - NVIDIA GeForce RTX 3070 (NVIDIA; 32.0.15.6094) - 11th Gen Intel(R) Core(TM) i7-11700K @ 3.60GHz (16 Threads)

### Issue description

The documentation for AudioEffectPitchShift claims that "1.0 is the default pitch and plays sounds unaffected", which matches expectations, however this is not the case. If you first set the pitch to something other than 1 then back to 1 via scripting, the sound will be noticeably distorted compared to how it initially was. When starting the effect at a pitch of 1 the audio is indeed unaffected, however changing it to something else then back to 1 continues to distort the sound.

In the video below, the pitch starts at a default 1.0, then I quickly press buttons to set the pitch to something else then back to 1.0, check the console to see the values. The sound is initially undistorted, but after changing its pitch to something else then back, there's an audible difference in the sound.

https://github.com/user-attachments/assets/a266e974-c495-442c-a940-2ba32eefca89

I was hoping to fade the pitch effect in and out as you open a menu, however this won't be possible without noticeable audio weirdness. I tried turning off the effect which works but has a noticeable snap to it as the effect instantly turns off.

### Steps to reproduce

- In the MRP below, you can press left / right arrows to set the pitch to 0.5 or 1.5. Pressing space returns the pitch to the default value of 1.0
- Enter play mode, listen to the sound
- Quickly press right arrow then space, the sound is noticeably not the same as how it started, despite the pitch being returned to the starting value of 1.0

### Minimal reproduction project (MRP)

[pitch-shift-bug.zip](https://github.com/user-attachments/files/16844151/pitch-shift-bug.zip)

### PR DESCRIPTION
**Pull request by pattlebass**
Fixes #96510

I believe the issue was some kind of `double` to `float` precision error.
It only occurred when changing the `pitch_scale` to a value different from `1.0` and then back to `1.0` (from GDScript or the inspector), but not by default or when hard-coding it to `1.0` in C++.

The project I used for testing:
[pitch-shift-bug.zip](https://github.com/user-attachments/files/17027217/pitch-shift-bug.zip)
Toggling the effect on and off should no longer yield any audible difference.

### CODE REVIEW

File: servers/audio/effects/audio_effect_pitch_shift.cpp [hunk 1]
```diff
@@ -286,6 +286,11 @@ void SMBPitchShift::smbFft(float *fftBuffer, long fftFrameSize, long sign)
 /* clang-format on */
 
 void AudioEffectPitchShiftInstance::process(const AudioFrame *p_src_frames, AudioFrame *p_dst_frames, int p_frame_count) {
+	// Avoid distortion by skipping processing if pitch_scale is 1.0
+	if (Math::is_equal_approx(base->pitch_scale, (float)1.0)) {
```
Comment (akien-mga) [inline servers/audio/effects/audio_effect_pitch_shift.cpp:290]:
```suggestion
	// Avoid distortion by skipping processing if pitch_scale is 1.0.
	if (Math::is_equal_approx(base->pitch_scale, 1.0f)) {
```

**Review by akien-mga — APPROVED** [review]

**Review by adamscott — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by JUSTCAMH** [comment]
I believe this fix won't have the intended effect. If you were to setup a script to linearly transition from pitch shift 1.1f to 1.0f, then you'd hear audible popping as the pitch shift effect turns itself off at pitch = 1.0. When I submitted the attached issue, I tried manually turning off the pitch shift effect when the pitch is 1.0, and this had a very noticeable popping to it.

We should fix the core problem (of this supposed double to float precision error) so that you can smoothly transition pitches without any popping or distortion.

**Comment by 3lis4** [comment]
Thank you all for your time on this issue.
@JUSTCAMH Do you know if it's planned somewhere to fix the core issue ?

This issue exists since a long time and wasn't properly fixed, see #20198. They also thought about disabling the effect for 1.0 value, but call it "a cheap workaround" : 
>  As soon as you activate the effect audio gets processed by an additional algorithm which transforms the original sound; regardless of the effect’s configuration, the input is changed.
> A cheap workaround would be to bypass the effect as long as the value is 1.

It's probably good to have this workaround for now, but it shouldn't prevent from fixing the core issue later on.

**Pull request merged**